File: JavaScript/Recursion/array-union.js
// Similar to sum.js

function arrayUnion(array1 = [], array2 = [], ...rest) {
  const result = [...array1, ...array2];
  if (!rest.length) return result;
  return arrayUnion(result, ...rest);
}

const expected = ["a", "b", "c", "a", "b", "d", "e", "z", "x", "w", "y"];

const result = arrayUnion(
  ["a", "b", "c"],
  ["a", "b", "d", "e"],
  ["z"],
  ["x"],
  ["w"],
  ["y"]
);

console.log(expected);
console.log(result);

File: Elixir/Recursion/array-union.exs
defmodule Exercise do
  def array_union([]), do: []

  def array_union([array1 | rest]) when is_list(array1) and rest == [],
    do: array1

  def array_union([array1 | [array2 | rest]]) when is_list(array1) and is_list(array2),
    do: array_union([array1 ++ array2] ++ rest)

  def run() do
    expected = ["a", "b", "c", "a", "b", "d", "e", "z", "x", "w", "y"]

    result =
      array_union([
        ["a", "b", "c"],
        ["a", "b", "d", "e"],
        ["z"],
        ["x"],
        ["w"],
        ["y"]
      ])

    IO.inspect(expected)
    IO.inspect(result)
  end
end

Exercise.run()
